py3status/modules/vpn_status.py

```diff
@@ -0,0 +1,140 @@
+#!/usr/bin/env python3
+"""
+Drop-in replacement for i3status run_watch VPN module.
+
+Expands on the i3status module by displaying the name of the connected vpn
+using pydbus. Asynchronously updates on dbus signals unless check_pid is True.
+
+Configuration parameters:
+    cache_timeout: How often to refresh in seconds when check_pid is True.
+        (default 10)
+    check_pid: If True, act just like the default i3status module.
+        (default False)
+    format: Format of the output.
+        (default 'VPN: {name}')
+    pidfile: Same as i3status.conf pidfile, checked when check_pid is True.
+        (default '/sys/class/net/vpn0/dev_id')
+
+Format string parameters:
+    {name} The name and/or status of the VPN.
+
+Requires:
+    pydbus: Which further requires PyGi. Check your distribution's repositories.
+
+@author Nathan Smith <nathan AT praisetopia.org>
+"""
+
+from pydbus import SystemBus
+from gi.repository import GObject
+from threading import Thread
+from os import path
+from time import time, sleep
+
+
+class Py3status:
+    # Available Configuration Parameters
+    cache_timeout = 10
+    check_pid = False
+    format = "VPN: {name}"
+    pidfile = '/sys/class/net/vpn0/dev_id'
+
+    def __init__(self):
+        self.thread_started = False
+        self.active = []
+
+    def _start_handler_thread(self):
+        """Called once to start the event handler thread."""
+        # Create handler thread
+        t = Thread(target=self._start_loop)
+        t.daemon = True
+
+        # Start handler thread
+        t.start()
+        self.thread_started = True
+
+    def _start_loop(self):
+        """Starts main event handler loop, run in handler thread t."""
+        # Create our main loop, get our bus, and add the signal handler
+        loop = GObject.MainLoop()
+        bus = SystemBus()
+        manager = bus.get(".NetworkManager")
+        manager.onPropertiesChanged = self._vpn_signal_handler
+
+        # Loop forever
+        loop.run()
+
+    def _vpn_signal_handler(self, args):
+        """Called on NetworkManager PropertiesChanged signal"""
+        # Args is a dictionary of changed properties
+        # We only care about changes in ActiveConnections
+        active = "ActiveConnections"
+        # Compare current ActiveConnections to last seen ActiveConnections
+        if active in args.keys() and sorted(self.active) != sorted(args[active]):
+            self.py3.update()
+            self.active = args[active]
```

Comment:
set `self.active` before calling `update()` since you want active set when you do the update